src/main/java/sorald/event/models/miner/MinedRule.java

```diff
@@ -0,0 +1,29 @@
+package sorald.event.models.miner;
+
+import com.google.common.collect.ImmutableList;
+import sorald.event.models.WarningLocation;
+
+public class MinedRule {
+    private final String ruleKey;
+    private final String ruleName;
+    private final ImmutableList<WarningLocation> warningLocations;
+
+    public MinedRule(
+            String ruleKey, String ruleName, ImmutableList<WarningLocation> warningLocations) {
+        this.ruleKey = ruleKey;
+        this.ruleName = ruleName;
+        this.warningLocations = warningLocations;
+    }
```

Comment:
Typically, what I do when I want to have immutable lists is to use standard library stuff. The input doesn't need to be immutable, either, it can be any list.

```suggestion
    public MinedRule(
            String ruleKey, String ruleName, List<WarningLocation> warningLocations) {
        this.ruleKey = ruleKey;
        this.ruleName = ruleName;
        this.warningLocations = Collections.unmodifiableList(warningLocations);
    }
```

Reply:
Caveat: `Collections.unmodifiableList` provides a read-only view of a list, so if you want to make sure the list isn't modified elsewhere, it also needs to be copied. E.g. `Collections.unmodifiableList(new ArrayList<>(warningLocations))`